### ISSUE TITLE
MoveXY and Move: A Plea for a Consistant API

### ISSUE THREAD

**Issue by AaronJRubin** [codecombat/codecombat#2907]
I was recently delighted to learn about the `move` method for incremental motion towards a target provided by various high-level footwear, which prevented me from needed to implement it myself on every level that it was useful (most of them). However, there are issues that I have with the naming conventions and parameters for `move` and `moveXY` that I think serve as an unnecessary barrier for people learning to code with this excellent game.

1. The fact that `move` refers to moving incrementally and `moveXY` refers to moving until a point is reached is deeply unintuitive. There's nothing about the naming of these methods that would lead one to associate one behavior with one and the other behavior with the other. What about `moveTowardXY` and `moveToXY`? It's more to type, but with the code completion facilities that the editor provides, that's not really a bit deal.

2. I see no good reason for these two methods to use different calling conventions. Why should `move` take a dict/object with x and y fields, but `moveXY` take two parameters? I can sort of see the logic - the fact that the name includes `XY` suggests an X parameter and a Y parameter - but that's completely ancillary to the actual difference in behavior of the two methods. It seems very arbitrary, and we should be showing new programmers that programming is about making abstractions clean and intuitive, not about memorizing arbitrary rules. Let them deal with messy API's when and if they have to learn OpenGL. :smile:

2. Even if there was a good reason for the two methods to use different calling conventions, differences between programming language idioms make the current use of a dict/object for `move` problematic. In Python, there are two idiomatic ways that one might represent the Javascript idiom of passing in an object literal with x and y properties:

  1. Use named parameters, so the call would be like `self.move(x = 50, y = 25)`
  2. Use an object/namedtuple, so the call would be like `self.move(Pos(50, 25))`. This is what we'd probably want to do here, since the reason for the use of a Javascript object literal is to stand in place of a pos object.

But passing in a dict with x and y fields in lieu of a pos object is not how things are normally done in Python, and for good reason - if the method expects an object, its fields would be accessed with dot syntax, which doesn't work with Python dicts! Sure, you could include a runtime check for whether the thing passed in is an instance of dict, and in that case use dict accessing syntax instead of object property accessing syntax… but that's pretty hacky.

In summary, my suggestions are:

1. Rename `move` to `moveTowardXY` and `moveXY` to `moveToXY`
2. Have both methods take two positional arguments, one for x and one for y.

**Comment by nwinter** [comment]
I agree with your reasoning and vaguely remember intending to have something like `moveToXY`, `moveToward`, `moveToXY`, and `moveToward`, where the XY versions are used earlier in the game when explaining more complicated syntax would be a burden, transitioning cleanly to the one-argument versions when object literals are introduced. I either forgot to rename things like that or had some other reason I've now forgotten not to do it.

Then people persuaded me that the higher-level boots should have both `move` and `moveXY` on them instead of just `move` so that they don't error out when playing the earlier levels, mostly. So if we kept that idea and expanded and renamed now, we'd end up with all four versions. I guess we can get rid of `moveTowardXY` if we want to save one, since incremental moving is taught after object literals.

I don't want to just use the XY version, since that makes things like `self.move(item.pos)` harder than it needs to be. And we do introduce the `Vector` class in the glacier, where the player is making her own Vectors and moving to them, so it's also less convenient to split a Vector into two arguments again just to move toward it there.

The Vector solution doesn't help too much with the non-Pythonicity of using a dict, though, because we don't have time to introduce Vectors until deep into the game.

All told, we were doing it again from scratch, I'd of course want to do it differently, but the workable improvement (making `moveToward`, `moveToXY`, and `moveTo`) seems messy enough that I don't think it's worth the trouble of migrating everything.
